File: tests/python/integration_test.py
"""
Python SDK Integration Tests

Validates blockchain transactions, user management, and cross-SDK compatibility
"""

import asyncio
import time
import uuid
from dataclasses import dataclass, field
from typing import Dict, List, Optional
from enum import Enum


class TransactionType(Enum):
    """Transaction types supported across all SDKs"""

    REGISTER_USER = "RegisterUser"
    SEND_DIRECT_MESSAGE = "SendDirectMessage"
    CREATE_CHANNEL = "CreateChannel"
    POST_TO_CHANNEL = "PostToChannel"
    VOTE_ON_GOVERNANCE = "VoteOnGovernance"


class TransactionStatus(Enum):
    """Transaction status enum"""

    PENDING = "Pending"
    CONFIRMED = "Confirmed"
    FAILED = "Failed"


@dataclass
class Transaction:
    """Transaction data structure"""

    tx_id: str
    tx_type: TransactionType
    sender: str
    data: Dict[str, str]
    timestamp: str
    status: TransactionStatus
    confirmations: int
    block_height: int = 1


class MockBlockchainPython:
    """Mock blockchain for testing"""

    def __init__(self, confirmation_threshold: int = 6):
        self.transactions: Dict[str, Transaction] = {}
        self.current_block: int = 1
        self.confirmation_threshold = confirmation_threshold

    def submit_transaction(
        self,
        tx_type: TransactionType,
        sender: str,
        data: Dict[str, str],
    ) -> str:
        """Submit a transaction"""
        tx_id = str(uuid.uuid4())
        tx = Transaction(
            tx_id=tx_id,
            tx_type=tx_type,
            sender=sender,
            data=data,
            timestamp="2025-10-29T10:00:00Z",
            status=TransactionStatus.CONFIRMED,
            confirmations=self.confirmation_threshold,
            block_height=self.current_block,
        )
        self.transactions[tx_id] = tx
        return tx_id

    def get_transaction(self, tx_id: str) -> Optional[Transaction]:
        """Get transaction by ID"""
        return self.transactions.get(tx_id)

    def get_all_transactions(self) -> List[Transaction]:
        """Get all transactions"""
        return list(self.transactions.values())

    def get_transactions_by_type(self, tx_type: TransactionType) -> List[Transaction]:
        """Get transactions by type"""
        return [tx for tx in self.transactions.values() if tx.tx_type == tx_type]

    def get_transactions_by_sender(self, sender: str) -> List[Transaction]:
        """Get transactions by sender"""
        return [tx for tx in self.transactions.values() if tx.sender == sender]

    def advance_blocks(self, count: int) -> None:
        """Advance blocks"""
        self.current_block += count
        for tx in self.transactions.values():
            if tx.status == TransactionStatus.CONFIRMED:
                tx.confirmations += count

    def get_current_block(self) -> int:
        """Get current block height"""
        return self.current_block

    def reset(self) -> None:
        """Reset blockchain state"""
        self.transactions.clear()
        self.current_block = 1


# ===== TESTS =====


async def test_user_registration():
    """Test user registration transaction"""
    blockchain = MockBlockchainPython()

    tx_id = blockchain.submit_transaction(
        TransactionType.REGISTER_USER,
        "alice",
        {"user_id": "alice-12345", "public_key": "alice-pub-key"},
    )

    tx = blockchain.get_transaction(tx_id)
    assert tx is not None
    assert tx.tx_type == TransactionType.REGISTER_USER
    assert tx.sender == "alice"
    assert tx.status == TransactionStatus.CONFIRMED
    print("✓ User registration test passed")


async def test_direct_message():
    """Test direct message transaction"""
    blockchain = MockBlockchainPython()

    # Create users
    blockchain.submit_transaction(TransactionType.REGISTER_USER, "alice", {"user_id": "alice-1"})
    blockchain.submit_transaction(TransactionType.REGISTER_USER, "bob", {"user_id": "bob-1"})

    # Send message
    msg_tx_id = blockchain.submit_transaction(
        TransactionType.SEND_DIRECT_MESSAGE,
        "alice",
        {"recipient_id": "bob-1", "content_hash": "msg-hash-123"},
    )

    tx = blockchain.get_transaction(msg_tx_id)
    assert tx is not None
    assert tx.tx_type == TransactionType.SEND_DIRECT_MESSAGE
    assert tx.sender == "alice"
    print("✓ Direct message test passed")


async def test_channel_creation():
    """Test channel creation transaction"""
    blockchain = MockBlockchainPython()

    # Create user
    blockchain.submit_transaction(TransactionType.REGISTER_USER, "alice", {"user_id": "alice-1"})

    # Create channel
    channel_tx_id = blockchain.submit_transaction(
        TransactionType.CREATE_CHANNEL,
        "alice",
        {"channel_name": "general", "description": "General discussion"},
    )

    tx = blockchain.get_transaction(channel_tx_id)
    assert tx is not None
    assert tx.tx_type == TransactionType.CREATE_CHANNEL
    assert tx.sender == "alice"
    print("✓ Channel creation test passed")


async def test_channel_posting():
    """Test posting to channel"""
    blockchain = MockBlockchainPython()

    # Create user
    blockchain.submit_transaction(TransactionType.REGISTER_USER, "alice", {"user_id": "alice-1"})

    # Create channel
    blockchain.submit_transaction(
        TransactionType.CREATE_CHANNEL, "alice", {"channel_name": "general"}
    )

    # Post to channel
    post_tx_id = blockchain.submit_transaction(
        TransactionType.POST_TO_CHANNEL,
        "alice",
        {"channel_id": "general", "content_hash": "post-hash-123"},
    )

    tx = blockchain.get_transaction(post_tx_id)
    assert tx is not None
    assert tx.tx_type == TransactionType.POST_TO_CHANNEL
    assert tx.sender == "alice"
    print("✓ Channel posting test passed")


async def test_transaction_filtering():
    """Test transaction filtering by type and sender"""
    blockchain = MockBlockchainPython()

    # Create mix
    blockchain.submit_transaction(TransactionType.REGISTER_USER, "alice", {"user_id": "alice-1"})
    blockchain.submit_transaction(TransactionType.REGISTER_USER, "bob", {"user_id": "bob-1"})
    blockchain.submit_transaction(
        TransactionType.CREATE_CHANNEL, "alice", {"channel_name": "general"}
    )

    # Filter by type
    reg_txs = blockchain.get_transactions_by_type(TransactionType.REGISTER_USER)
    assert len(reg_txs) == 2

    channel_txs = blockchain.get_transactions_by_type(TransactionType.CREATE_CHANNEL)
    assert len(channel_txs) == 1

    # Filter by sender
    alice_txs = blockchain.get_transactions_by_sender("alice")
    assert len(alice_txs) == 2

    bob_txs = blockchain.get_transactions_by_sender("bob")
    assert len(bob_txs) == 1

    print("✓ Transaction filtering test passed")


async def test_user_activity_history():
    """Test tracking user activity history"""
    blockchain = MockBlockchainPython()

    # User performs activities
    activities = [
        (TransactionType.REGISTER_USER, "Register"),
        (TransactionType.SEND_DIRECT_MESSAGE, "Message"),
        (TransactionType.CREATE_CHANNEL, "Channel"),
        (TransactionType.POST_TO_CHANNEL, "Post"),
    ]

    for tx_type, label in activities:
        blockchain.submit_transaction(tx_type, "alice", {"content": label})

    alice_txs = blockchain.get_transactions_by_sender("alice")
    assert len(alice_txs) == 4
    print("✓ User activity history test passed")


async def test_confirmation_tracking():
    """Test confirmation tracking"""
    blockchain = MockBlockchainPython(confirmation_threshold=6)

    tx_id = blockchain.submit_transaction(
        TransactionType.REGISTER_USER, "alice", {"user_id": "alice-1"}
    )

    tx = blockchain.get_transaction(tx_id)
    assert tx.confirmations == 6

    blockchain.advance_blocks(5)
    tx = blockchain.get_transaction(tx_id)
    assert tx.confirmations == 11
    print("✓ Confirmation tracking test passed")


async def test_block_height_tracking():
    """Test block height tracking"""
    blockchain = MockBlockchainPython()

    assert blockchain.get_current_block() == 1

    blockchain.submit_transaction(TransactionType.REGISTER_USER, "alice", {"user_id": "alice-1"})
    blockchain.advance_blocks(5)

    assert blockchain.get_current_block() == 6

    blockchain.submit_transaction(TransactionType.REGISTER_USER, "bob", {"user_id": "bob-1"})
    tx = blockchain.get_transactions_by_sender("bob")[0]
    assert tx.block_height == 6
    print("✓ Block height tracking test passed")


async def test_concurrent_operations():
    """Test concurrent user operations"""
    blockchain = MockBlockchainPython()
    users = ["alice", "bob", "charlie"]

    # Register all users
    for user in users:
        blockchain.submit_transaction(
            TransactionType.REGISTER_USER, user, {"user_id": f"{user}-id"}
        )

    # Each creates a channel
    for user in users:
        blockchain.submit_transaction(
            TransactionType.CREATE_CHANNEL, user, {"channel_name": f"{user}-channel"}
        )

    # Each sends a message
    for i, user in enumerate(users):
        blockchain.submit_transaction(
            TransactionType.SEND_DIRECT_MESSAGE,
            user,
            {"recipient": users[(i + 1) % len(users)]},
        )

    stats_by_user = {user: len(blockchain.get_transactions_by_sender(user)) for user in users}
    for user in users:
        assert stats_by_user[user] == 3  # 1 registration + 1 channel + 1 message

    print("✓ Concurrent operations test passed")


async def run_tests():
    """Run all tests"""
    print("\n📋 Python SDK Integration Tests\n")
    print("Running 9 test cases...\n")

    await test_user_registration()
    await test_direct_message()
    await test_channel_creation()
    await test_channel_posting()
    await test_transaction_filtering()
    await test_user_activity_history()
    await test_confirmation_tracking()
    await test_block_height_tracking()
    await test_concurrent_operations()

    print("\n✅ All Python integration tests passed!\n")


if __name__ == "__main__":
    asyncio.run(run_tests())

File: tests/typescript/integration.test.ts
// Integration Tests for TypeScript SDK
// Validates blockchain and user management operations

import { strict as assert } from "assert";
// Note: uuid package must be installed: npm install uuid
const uuidv4 = () => Math.random().toString(36).substring(2, 15);

interface Transaction {
  tx_id: string;
  tx_type: string;
  sender: string;
  data: Record<string, string>;
  timestamp: string;
  status: "Pending" | "Confirmed" | "Failed";
  confirmations: number;
}

interface User {
  user_id: string;
  username: string;
  public_key: string;
  on_chain_confirmed: boolean;
}

class MockBlockchainTS {
  private transactions: Map<string, Transaction> = new Map();
  private currentBlock: number = 1;
  private confirmationThreshold: number = 6;

  submitTransaction(
    txType: string,
    sender: string,
    data: Record<string, string>
  ): string {
    const txId = uuidv4();
    const tx: Transaction = {
      tx_id: txId,
      tx_type: txType,
      sender,
      data,
      timestamp: new Date().toISOString(),
      status: "Confirmed",
      confirmations: this.confirmationThreshold,
    };

    this.transactions.set(txId, tx);
    return txId;
  }

  getTransaction(txId: string): Transaction | undefined {
    return this.transactions.get(txId);
  }

  getTransactionsByType(txType: string): Transaction[] {
    return Array.from(this.transactions.values()).filter(
      (tx) => tx.tx_type === txType
    );
  }

  getTransactionsBySender(sender: string): Transaction[] {
    return Array.from(this.transactions.values()).filter(
      (tx) => tx.sender === sender
    );
  }

  getAllTransactions(): Transaction[] {
    return Array.from(this.transactions.values());
  }

  getCurrentBlock(): number {
    return this.currentBlock;
  }

  advanceBlocks(count: number): void {
    this.currentBlock += count;
    // Update confirmations
    for (const tx of this.transactions.values()) {
      if (tx.status === "Confirmed") {
        tx.confirmations += count;
      }
    }
  }

  reset(): void {
    this.transactions.clear();
    this.currentBlock = 1;
  }
}

// ===== TESTS =====

async function testUserRegistrationTransaction() {
  const blockchain = new MockBlockchainTS();

  const txId = blockchain.submitTransaction(
    "RegisterUser",
    "alice",
    {
      user_id: "alice-12345",
      public_key: "alice-pub-key",
    }
  );

  const tx = blockchain.getTransaction(txId);
  assert(tx !== undefined);
  assert.equal(tx.tx_type, "RegisterUser");
  assert.equal(tx.sender, "alice");
  assert.equal(tx.status, "Confirmed");

  console.log("✓ User registration transaction test passed");
}

async function testDirectMessageTransaction() {
  const blockchain = new MockBlockchainTS();

  // Create users
  blockchain.submitTransaction("RegisterUser", "alice", {
    user_id: "alice-1",
  });
  blockchain.submitTransaction("RegisterUser", "bob", {
    user_id: "bob-1",
  });

  // Send message
  const msgTxId = blockchain.submitTransaction(
    "SendDirectMessage",
    "alice",
    {
      recipient_id: "bob-1",
      content_hash: "msg-hash-123",
    }
  );

  const tx = blockchain.getTransaction(msgTxId);
  assert(tx !== undefined);
  assert.equal(tx.tx_type, "SendDirectMessage");
  assert.equal(tx.sender, "alice");

  console.log("✓ Direct message transaction test passed");
}

async function testChannelCreationTransaction() {
  const blockchain = new MockBlockchainTS();

  // Create user
  blockchain.submitTransaction("RegisterUser", "alice", {
    user_id: "alice-1",
  });

  // Create channel
  const channelTxId = blockchain.submitTransaction(
    "CreateChannel",
    "alice",
    {
      channel_name: "general",
      description: "General discussion",
    }
  );

  const tx = blockchain.getTransaction(channelTxId);
  assert(tx !== undefined);
  assert.equal(tx.tx_type, "CreateChannel");
  assert.equal(tx.sender, "alice");

  console.log("✓ Channel creation transaction test passed");
}

async function testTransactionFilteringByType() {
  const blockchain = new MockBlockchainTS();

  // Create mix
  blockchain.submitTransaction("RegisterUser", "alice", { user_id: "alice-1" });
  blockchain.submitTransaction("RegisterUser", "bob", { user_id: "bob-1" });
  blockchain.submitTransaction("CreateChannel", "alice", {
    channel_name: "general",
  });

  const regTxs = blockchain.getTransactionsByType("RegisterUser");
  assert.equal(regTxs.length, 2);

  const channelTxs = blockchain.getTransactionsByType("CreateChannel");
  assert.equal(channelTxs.length, 1);

  console.log("✓ Transaction filtering by type test passed");
}

async function testTransactionFilteringBySender() {
  const blockchain = new MockBlockchainTS();

  blockchain.submitTransaction("RegisterUser", "alice", { user_id: "alice-1" });
  blockchain.submitTransaction("CreateChannel", "alice", {
    channel_name: "general",
  });
  blockchain.submitTransaction("RegisterUser", "bob", { user_id: "bob-1" });

  const aliceTxs = blockchain.getTransactionsBySender("alice");
  assert.equal(aliceTxs.length, 2);

  const bobTxs = blockchain.getTransactionsBySender("bob");
  assert.equal(bobTxs.length, 1);

  console.log("✓ Transaction filtering by sender test passed");
}

async function testMultipleUsersMessageFlow() {
  const blockchain = new MockBlockchainTS();
  const users = ["alice", "bob", "charlie"];

  // Register users
  for (const user of users) {
    blockchain.submitTransaction("RegisterUser", user, { user_id: `${user}-id` });
  }

  // Exchange messages
  for (const user of users) {
    blockchain.submitTransaction("SendDirectMessage", user, {
      recipient_id: "bob-id",
      content: `Message from ${user}`,
    });
  }

  assert.equal(blockchain.getAllTransactions().length, 6); // 3 registrations + 3 messages

  for (const user of users) {
    const txs = blockchain.getTransactionsBySender(user);
    assert.equal(txs.length, 2); // 1 registration + 1 message
  }

  console.log("✓ Multiple users message flow test passed");
}

async function testBlockHeightTracking() {
  const blockchain = new MockBlockchainTS();

  assert.equal(blockchain.getCurrentBlock(), 1);

  blockchain.submitTransaction("RegisterUser", "alice", { user_id: "alice-1" });
  blockchain.advanceBlocks(5);

  assert.equal(blockchain.getCurrentBlock(), 6);

  console.log("✓ Block height tracking test passed");
}

// ===== RUN TESTS =====

async function runTests() {
  console.log("\n📋 TypeScript SDK Integration Tests\n");
  console.log("Running 7 test cases...\n");

  await testUserRegistrationTransaction();
  await testDirectMessageTransaction();
  await testChannelCreationTransaction();
  await testTransactionFilteringByType();
  await testTransactionFilteringBySender();
  await testMultipleUsersMessageFlow();
  await testBlockHeightTracking();

  console.log("\n✅ All TypeScript integration tests passed!\n");
}

runTests().catch(console.error);
